scanner.go

```diff
@@ -156,62 +161,96 @@ func newScanner(c RPCClient, rpc *hrpc.Scan, logger *slog.Logger) *scanner {
 	}
 }
 
-func toLocalResult(r *pb.Result) *hrpc.Result {
-	if r == nil {
-		return nil
+func newScannerV2(c RPCClient, rpc *hrpc.Scan, logger *slog.Logger) *scannerV2 {
+	var sm map[string]int64
+	if rpc.TrackScanMetrics() {
+		sm = make(map[string]int64)
+	}
+	return &scannerV2{
+		scanner: scanner{
+			RPCClient:          c,
+			rpc:                rpc,
+			startRow:           rpc.StartRow(),
+			curRegionScannerID: noScannerID,
+			scanMetrics:        sm,
+			logger:             logger,
+		},
 	}
-	res := hrpc.ToLocalResult(r)
-	return res
 }
 
 func (s *scanner) Next() (*hrpc.Result, error) {
-	var (
-		result, partial *pb.Result
-		err             error
-	)
-
-	select {
-	case <-s.rpc.Context().Done():
-		s.Close()
-		return nil, s.rpc.Context().Err()
-	default:
+	if err := s.rpc.Context().Err(); err != nil {
+		return nil, err
 	}
 
 	if s.rpc.AllowPartialResults() {
 		// if client handles partials, just return it
-		result, err = s.peek()
+		result, err := s.peek()
 		if err != nil {
 			return nil, err
 		}
 		s.shift()
-		return toLocalResult(result), nil
+		return hrpc.ResultV2ToResult(result), nil
 	}
 
+	var (
+		result, partial hrpc.ResultV2
+		err             error
+	)
+
 	for {
 		partial, err = s.peek()
-		if err == io.EOF && result != nil {
+		if err == io.EOF && len(result.Cells) > 0 {
 			// no more results, return what we have. Next call to the Next() will get EOF
-			result.Partial = proto.Bool(false)
-			return toLocalResult(result), nil
+			result.Partial = false
+			return hrpc.ResultV2ToResult(result), nil
 		}
 		if err != nil {
 			// return whatever we have so far and the error
-			return toLocalResult(result), err
+			return hrpc.ResultV2ToResult(result), err
 		}
 
 		var done bool
 		result, done = s.coalesce(result, partial)
 		if done {
 			s.shift()
 		}
-		if !result.GetPartial() {
+		if !result.Partial {
 			// if not partial anymore, return it
-			return toLocalResult(result), nil
+			return hrpc.ResultV2ToResult(result), nil
 		}
 	}
 }
 
-func (s *scanner) request() (*pb.ScanResponse, hrpc.RegionInfo, error) {
+func (s *scannerV2) Next() (*hrpc.ScanResponseV2, error) {
+	if s.renewCancel != nil {
+		// About to send new Scan request to HBase, cancel our
+		// renewer.
+		s.renewCancel()
+		s.renewCancel = nil
+	}
+
+	if s.closed {
+		return nil, io.EOF
+	}
+
+	if err := s.rpc.Context().Err(); err != nil {
+		return nil, err
+	}
+	resp, err := s.fetch()
+	if err != nil {
+		return nil, err
+	}
+	if !s.closed && s.rpc.RenewInterval() > 0 {
+		// Start up a renewer
+		renewCtx, cancel := context.WithCancel(s.rpc.Context())
+		s.renewCancel = cancel
+		go s.renewLoop(renewCtx, s.startRow)
```

Comment:
nit: you could end up leaking renewLoop goroutines if somebody was using Next concurrently, maybe we should warn against this in the documentation